lib/go-tc/deliveryservice_requests.go

```diff
@@ -547,3 +625,372 @@ func (r RequestStatus) ValidTransition(to RequestStatus) error {
+	if dsr.LastEditedByID != nil {
+		downgraded.LastEditedByID = new(IDNoMod)
+		*downgraded.LastEditedByID = IDNoMod(*dsr.LastEditedByID)
+	}
+	downgraded.LastUpdated = TimeNoModFromTime(dsr.LastUpdated)
+	*downgraded.Status = dsr.Status
+	if dsr.Original != nil && dsr.Original.XMLID != nil {
+		downgraded.XMLID = new(string)
+		*downgraded.XMLID = *dsr.Original.XMLID
+	} else if dsr.Requested.XMLID != nil {
+		downgraded.XMLID = new(string)
+		*downgraded.XMLID = *dsr.Requested.XMLID
+	}
+	return downgraded
+}
+
+// String encodes the DeliveryServiceRequestV30 as a string, in the format
+// "DeliveryServiceRequestV30({{Property}}={{Value}}[, {{Property}}={{Value}}]+)".
+//
+// If a property is a pointer value, then its dereferenced value is used -
+// unless it's nil, in which case "<nil>" is used as the value. Original and
+// Requested are omitted, because of how large they are. Times are formatted in
+// RFC3339 format.
+func (dsr DeliveryServiceRequestV30) String() string {
+	var builder strings.Builder
+	builder.Write([]byte("DeliveryServiceRequestV30(Assignee="))
+	if dsr.Assignee != nil {
+		builder.WriteRune('"')
+		builder.WriteString(*dsr.Assignee)
+		builder.WriteRune('"')
+	} else {
+		builder.Write([]byte("<nil>"))
+	}
+	builder.Write([]byte(", AssigneeID="))
+	if dsr.AssigneeID != nil {
+		builder.WriteString(strconv.Itoa(*dsr.AssigneeID))
+	} else {
+		builder.Write([]byte("<nil>"))
+	}
+	builder.Write([]byte(`, Author="`))
+	builder.WriteString(dsr.Author)
+	builder.Write([]byte(`", AuthorID=`))
+	if dsr.AuthorID != nil {
+		builder.WriteString(strconv.Itoa(*dsr.AuthorID))
+	} else {
+		builder.Write([]byte("<nil>"))
+	}
+	builder.Write([]byte(`, ChangeType="`))
+	builder.WriteString(dsr.ChangeType.String())
+	builder.Write([]byte(`", CreatedAt=`))
+	builder.WriteString(dsr.CreatedAt.Format(time.RFC3339))
+	builder.Write([]byte(", ID="))
+	if dsr.ID != nil {
+		builder.WriteString(strconv.Itoa(*dsr.ID))
+	} else {
+		builder.Write([]byte("<nil>"))
+	}
+	builder.Write([]byte(`, LastEditedBy="`))
+	builder.WriteString(dsr.LastEditedBy)
+	builder.Write([]byte(`", LastEditedByID=`))
+	if dsr.LastEditedByID != nil {
+		builder.WriteString(strconv.Itoa(*dsr.LastEditedByID))
+	} else {
+		builder.Write([]byte("<nil>"))
+	}
+	builder.Write([]byte(`, LastUpdated=`))
+	builder.WriteString(dsr.LastUpdated.Format(time.RFC3339))
+	builder.Write([]byte(`, Status="`))
+	builder.WriteString(dsr.Status.String())
+	builder.Write([]byte(`")`))
+	return builder.String()
+}
+
+// Validate satisfies the
+// github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api.ParseValidator
+// interface.
+func (dsr *DeliveryServiceRequestV30) Validate(tx *sql.Tx) error {
+	if tx == nil {
+		return errors.New("unknown error")
+	}
+
+	fromStatus := RequestStatusDraft
+	if dsr.ID != nil && *dsr.ID > 0 {
+		if err := tx.QueryRow(`SELECT status FROM deliveryservice_request WHERE id=$1`, *dsr.ID).Scan(&fromStatus); err != nil {
+			return err
+		}
```

Comment:
The user should not be able to see SQL errors.

Reply:
Generally I agree. This was done to mimic the behavior of the old API versions, which do/did return SQL errors back to the user. So I think it'd be best to settle the above discussion before getting into this one, since depending on the resolution this might no longer be an issue, and/or they'll be changed in the same way at the same time.

Reply:
So should I do what I did above here, or keep the existing error message? Clients shouldn't be looking for specific error messages, but I know they do. Even our own API tests look for exact wording in some places.

Reply:
nvm, I've got that backwards. I think I'll just not use the `ParseValidator` interface.

Reply:
Our API tests looking for exact wording was before the TO API tests had access to the response status codes. No reason the wording should matter now that we've got those.